File: Data Structures & Algorithms/maximum-product-subarray/submission-5.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        res = nums[0]
        currMax, currMin = 1, 1
        for n in nums:
            currMax, currMin = max(n, n*currMax, n*currMin), min(n, n*currMax, n*currMin)
            res = max(res, currMax)
        return res

File: Data Structures & Algorithms/burst-balloons/submission-5.cpp
class Solution {
public:
    int maxCoins(vector<int>& nums) {
        nums.insert(nums.begin(), 1);
        nums.push_back(1);
        int n = nums.size();
        vector<vector<int>> dp(n, vector<int>(n, 0));

        for (int len = 2; len < n; len++) {
            for (int l = 0; l + len < n; l++) {
                int r = l + len;
                for (int i = l+1; i < r; i++) {
                    dp[l][r] = max(dp[l][r], nums[l]*nums[i]*nums[r] + dp[l][i] + dp[i][r]);
                }
            }
        }
        return dp[0][n-1];
    }
};